lxd/backup.go

```diff
@@ -0,0 +1,297 @@
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/pkg/errors"
+)
+
+var ErrMissingIndexFile = errors.New("missing index.yaml")
+
+// backup represents a container backup.
+type backup struct {
+	state     *state.State
+	container container
+
+	// Properties
+	id               int
+	name             string
+	creationDate     time.Time
+	expiryDate       time.Time
+	containerOnly    bool
+	optimizedStorage bool
+}
+
+type backupInfo struct {
+	Name            string   `json:"name" yaml:"name"`
+	Backend         string   `json:"backend" yaml:"backend"`
+	Privileged      bool     `json:"privileged" yaml:"privileged"`
+	Pool            string   `json:"pool" yaml:"pool"`
+	Snapshots       []string `json:"snapshots,omitempty" yaml:"snapshots,omitempty"`
+	HasBinaryFormat bool     `json:"-" yaml:"-"`
+}
+
+// Rename renames a container backup.
+func (b *backup) Rename(newName string) error {
+	ourStart, err := b.container.StorageStart()
+	if err != nil {
+		return err
+	}
+	if ourStart {
+		defer b.container.StorageStop()
+	}
+
+	// Rename the database entry
+	err = b.state.Cluster.ContainerBackupRename(b.Name(), newName)
+	if err != nil {
+		return err
+	}
+
+	// Rename the directories and files
+	err = b.container.Storage().ContainerBackupRename(*b, newName)
+	if err != nil {
+		return err
+	}
```

Comment:
I mean, if we don't have time to do it right now we should still note that we should revert on error otherwise you'd end up with a mismatch between db and on-disk naming.

Reply:
And order wise, I thought we usually did FS before DB as the filesystem is far more likely to fail than the DB, so this should be re-ordered.

Reply:
fixed